### PR TITLE
Add notification scheme setter

### PR DESCRIPTION
**Pull request by Humbedooh**
To be documented :)
This adds notification scheme settings via .asf.yaml for repositories.
They are validated, then saved to a yaml file on disk.
To-Do: Use these settings...

### CODE REVIEW

File: modules/gitbox/files/asfgit/asfyaml.py [hunk 1]
```diff
@@ -406,3 +407,79 @@ def publish(cfg, yml):
...
+        'pullrequests_status',
+        'pullrequests_comment',
+    ]
+
+    # Get branch
+    ref = yml.get('refname', 'master').replace('refs/heads/', '')
+
+    # Ensure this is master, or bail
+    if ref != 'master' and ref != 'trunk':
+        print("[NOTICE] Notification scheme settings can only be applied to the master or trunk branch.")
```
Comment (gstein) [inline modules/gitbox/files/asfgit/asfyaml.py:429]: the PMC won't actually see this notice, though. ?

File: modules/gitbox/files/asfgit/asfyaml.py [hunk 2]
```diff
@@ -406,3 +407,79 @@ def publish(cfg, yml):
...
+        return
+
+    # infer project name
+    m = re.match(r"(?:incubator-)?([^-.]+)", cfg.repo_name)
+    pname = m.group(1)
```
Comment (gstein) [inline modules/gitbox/files/asfgit/asfyaml.py:434]: are all repo names proper? that regex will always match? ie. m != None

should we have something like `asfpy.util.extract_pmc(repo_name)` ? ... and that holds `WSMAP` and other related bits, for all ASF python code?

File: modules/gitbox/files/asfgit/asfyaml.py [hunk 3]
```diff
@@ -406,3 +407,79 @@ def publish(cfg, yml):
     except Exception as e:
         print(e)
         asfgit.log.exception()
+
+
+def notifications(cfg, yml):
+    """ Notification scheme setup """
+    valid_schemes = [
```
Comment (gstein) [inline modules/gitbox/files/asfgit/asfyaml.py:414]: move this magic value to GLOBAL_CONST_VALUE

File: modules/gitbox/files/asfgit/asfyaml.py [hunk 4]
```diff
@@ -406,3 +407,79 @@ def publish(cfg, yml):
...
+        # Verify that all set schemes pass muster and point to $foo@$project.a.o
+        if not re.match(r"[-a-z0-9]+@[-a-z0-9]+(\.incubator)?\.apache\.org$", v)\
+            or not (
+                v.endswith('@%s.apache.org' % pname) or
+                v.endswith('@%s.incubator.apache.org' % pname)
+            ):
+            raise Exception("Invalid notification target '%s'. Must be a valid @%s.apache.org list!" % (v, pname))
+
+    # All seems kosher, update settings if need be
+    scheme_path = os.path.join(cfg.repo_dir, 'notification_schemes.yaml')
```
Comment (gstein) [inline modules/gitbox/files/asfgit/asfyaml.py:455]: another magic value to move

should this be in the root dir on the repo, or in $repo/.git/ ?

File: modules/gitbox/files/asfgit/asfyaml.py [hunk 5]
```diff
@@ -406,3 +407,79 @@ def publish(cfg, yml):
...
+    m = re.match(r"(?:incubator-)?([^-.]+)", cfg.repo_name)
+    pname = m.group(1)
+    pname = WSMAP.get(pname, pname)
+
+    # Verify that we know all settings in the yaml
+    if type(yml) is not dict:
+        raise Exception("Notification schemes must be simple 'key: value' pairs!")
+
+    for k, v in yml.items():
+        if type(v) is not str:
```
Comment (gstein) [inline modules/gitbox/files/asfgit/asfyaml.py:442]: `if not isinstance(v, str):`

File: modules/gitbox/files/asfgit/asfyaml.py [hunk 6]
```diff
@@ -406,3 +407,79 @@ def publish(cfg, yml):
...
+            raise Exception("Invalid value for setting '%s' - must be string value!" % k)
+        if k not in valid_schemes:
+            raise Exception("Invalid notification scheme '%s' detected, please remove it!" % k)
+        # Verify that all set schemes pass muster and point to $foo@$project.a.o
+        if not re.match(r"[-a-z0-9]+@[-a-z0-9]+(\.incubator)?\.apache\.org$", v)\
```
Comment (gstein) [inline modules/gitbox/files/asfgit/asfyaml.py:447]: mebbe a global:
`RE_VALID_EMAIL = re.compile(...)`

not sure if we need that elsewhere? (ie. should such be moved to asfpy?)

File: modules/gitbox/files/asfgit/asfyaml.py [hunk 7]
```diff
@@ -406,3 +407,79 @@ def publish(cfg, yml):
...
+    print("Updating notification schemes for repository: ")
+
+    # Figure out what changed since last
+    for key in valid_schemes:
+        if key not in old_yml and key in yml:
+            print("- adding new scheme (%s): %s" % (key, yml[key]))
+        elif key in old_yml and key not in yml:
+            print("- removing old scheme (%s) - was %s" % (key, old_yml[key]))
+        elif key in old_yml and key in yml and old_yml[key] != yml[key]:
+            print("- updating scheme %s: %s -> %s" % (key, old_yml[key], yml[key]))
```
Comment (gstein) [inline modules/gitbox/files/asfgit/asfyaml.py:473]: do these print() calls end up in a log file somewhere? ... and/or should these be gathered into the email which is sent below?

File: modules/gitbox/files/asfgit/asfyaml.py (hunk 1 above)
Comment (Humbedooh) [reply modules/gitbox/files/asfgit/asfyaml.py:429]: Correct, only the pusher will see it, that is intended.

File: modules/gitbox/files/asfgit/asfyaml.py (hunk 7 above)
Comment (Humbedooh) [reply modules/gitbox/files/asfgit/asfyaml.py:473]: They end up in the git response when you push. The PMC gets an email with the updated settings, but I'm open to making that email have the same style.

File: modules/gitbox/files/asfgit/asfyaml.py (hunk 2 above)
Comment (Humbedooh) [reply modules/gitbox/files/asfgit/asfyaml.py:437]: We could do that, yes.
I think that's outside the scope of this particular PR, but something worth pursuing :)

Comment (Humbedooh) [reply modules/gitbox/files/asfgit/asfyaml.py:439]: FWIW, it will always match. If it didn't, projects wouldn't have write access to them.

**Pull request merged**